### PR TITLE
Feat: upgrade outlines & support compatibility with the old version

### PR DESCRIPTION
**Pull request by gobraves**
## Motivation

#2276 
upgrade outlines & support compatibility with the old version

Due to vLLM limitations, outlines is restricted to versions >= 0.0.43 and < 0.1. If you update the outlines version in pyproject.toml and use pip install "sglang[all]", the actual version of outlines will still remain below 0.1. Additionally, using uv may result in version conflicts.

After discussion with @merrymercy  in #2276 ,  I'll  update the code first and address pyproject.toml later.

If you wish to use the new version of outlines, you can run:
pip install outlines==0.1.7 outlines-core==0.1.17 --no-deps.

## Checklist

- [ ] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [ ] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/references/contributor_guide.md).
- [ ] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

**Review by merrymercy — APPROVED** [review]

File: python/sglang/srt/constrained/outlines_backend.py [hunk 1]
```diff
@@ -152,7 +152,11 @@ def init_value_impl(self, key: Tuple[str, str]) -> OutlinesGrammar:
             raise ValueError(f"Invalid key_type: {key_type}")
 
         try:
-            guide = RegexGuide(regex, self.outlines_tokenizer)
+            # Prior to outlines 0.1.1, the RegexGuide class did not include the from_regex function.
+            if hasattr(RegexGuide, "from_regex"):
+                guide = RegexGuide.from_regex(regex, self.outlines_tokenizer)
+            else:
+                guide = RegexGuide(regex, self.outlines_tokenizer)
```
Comment (merrymercy) [inline python/sglang/srt/constrained/outlines_backend.py]:
```suggestion
            if hasattr(RegexGuide, "from_regex"):
                # outlines >= 0.1.1
                guide = RegexGuide.from_regex(regex, self.outlines_tokenizer)
            else:
                # outlines <= 0.0.46
                guide = RegexGuide(regex, self.outlines_tokenizer)
```

File: python/sglang/srt/constrained/outlines_jump_forward.py [hunk 2]
```diff
@@ -23,7 +23,13 @@
...
 from interegular import InvalidSyntax
 from outlines.caching import cache as disk_cache
-from outlines.fsm.regex import FSMInfo, make_byte_level_fsm, make_deterministic_fsm
+
+try:
+    from outlines_core.fsm.outlines_core_rs import FSMInfo
+    from outlines_core.fsm.regex import make_byte_level_fsm, make_deterministic_fsm
+except ImportError:
+    # outlines <= 0.0.46
+    from outlines.fsm.regex import FSMInfo, make_byte_level_fsm, make_deterministic_fsm
```
Comment (merrymercy) [inline python/sglang/srt/constrained/outlines_jump_forward.py]:
```suggestion
try:
    # outlines >= 0.1.0
    from outlines_core.fsm.outlines_core_rs import FSMInfo
    from outlines_core.fsm.regex import make_byte_level_fsm, make_deterministic_fsm
except ImportError:
    # outlines <= 0.0.46
    from outlines.fsm.regex import FSMInfo, make_byte_level_fsm, make_deterministic_fsm
```

**Comment by Swipe4057** [comment]
a warning appears in the logs:
/data/venvs/sglang/lib/python3.12/site-packages/sglang/srt/constrained/outlines_backend.py:88: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  tokens = torch.tensor()
  
Response time increases +69% if you use pip install outlines==0.1.7 outlines-core==0.1.17 --no-deps.
